adapters/repos/db/vector/hnsw/packedconn/connections.go

```diff
@@ -0,0 +1,327 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2023 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package packedconn
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+	"sort"
+)
+
+// Data order
+// | Start | Len        | Description |
+// | ----- | ---------- | ------------|
+// | 0     | 1          | uint8 indicating count of layers of object |
+// | 1     | 3 * layers | uint8 len indicator, followed by uint16 offset indicator |
+// | dyn.  | dyn        | all layers, starting with highest layer first
+
+type Connections struct {
+	data []byte
+	buff []byte
+}
+
+const (
+	layerPos    = 0
+	initialSize = 50
+)
+
+func NewWithMaxLayer(maxLayer uint8) (*Connections, error) {
+	if int(maxLayer)+1 > math.MaxUint8 {
+		return nil, fmt.Errorf("max supported layer is %d",
+			math.MaxUint8-1)
+	}
+	c := Connections{
+		// TODO: low initial size and grow dynamically
+		data: make([]byte, initialSize+1+(maxLayer+1)*3),
+	}
+
+	c.initLayers(maxLayer)
+
+	c.buff = make([]byte, 16)
+
+	return &c, nil
+}
+
+func (c *Connections) AddLayer() {
+	layers := c.Layers()
+	c.expandDataIfRequired(3)
+	c.shiftRightBy(1+uint16(layers)*3, 3)
+	c.data[layerPos] = layers + 1
+	c.setLayerLength(layers, 0)
+	c.setLayerOffset(layers, 1+uint16(layers+1)*3)
+	layer := layers - 1
+	for {
+		c.setLayerOffset(layer, c.layerOffset(layer)+3)
+
+		if layer == 0 {
+			break
+		}
+		layer--
+	}
+}
+
+func (c *Connections) ReplaceLayer(layer uint8, conns []uint64) {
+	// create a temporary buffer that is guaranteed to fit everything. The
+	// over-allocation does not matter, this buffer won't stick around, so the
+	// only real downside is the overhead on GC. If this because noticeable this
+	// buffer would be suitable to use pooling.
+	buf := make([]byte, len(conns)*binary.MaxVarintLen64)
+
+	sort.Slice(conns, func(a, b int) bool { return conns[a] < conns[b] })
+	last := uint64(0)
+	offset := 0
+	for _, raw := range conns {
+		delta := raw - last
+		last = raw
+		offset += binary.PutUvarint(buf[offset:], delta)
+	}
+
+	buf = buf[:offset]
+
+	c.replaceLayer(layer, buf, uint8(len(conns)))
+}
+
+func (c Connections) LenAtLayer(layer uint8) int {
+	if layer >= c.Layers() {
+		panic(fmt.Sprintf("only has %d layers", c.Layers()))
+	}
+
+	return int(c.layerLength(layer))
+}
+
+func (c Connections) CopyLayer(conns []uint64, layer uint8) []uint64 {
+	if cap(conns) < int(c.layerLength(layer)) {
```

Comment:
`c.layerLength(layer)` can be stored in variable, to avoid multiple calls and improve readibility